core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigMappingUtils.java

```diff
@@ -85,45 +87,76 @@ static void processConfigClass(
-                            mappingMetadata.getClassBytes()));
-            reflectiveClasses
-                    .produce(ReflectiveClassBuildItem.builder(mappingMetadata.getInterfaceType()).methods().build());
-            reflectiveClasses
-                    .produce(ReflectiveClassBuildItem.builder(mappingMetadata.getClassName())
-                            .methods().build());
-
+            generatedClassesNames.add(mappingMetadata.getClassName());
+            // This is the generated implementation of the mapping by SmallRye Config.
+            generatedClasses.produce(new GeneratedClassBuildItem(isApplicationClass, mappingMetadata.getClassName(),
+                    mappingMetadata.getClassBytes()));
+            // Register the interface and implementation methods for reflection. This is required for Bean Validation.
+            reflectiveClasses.produce(ReflectiveClassBuildItem.builder(mappingMetadata.getInterfaceType()).methods().build());
+            reflectiveClasses.produce(ReflectiveClassBuildItem.builder(mappingMetadata.getClassName()).methods().build());
+            // Register also the interface hierarchy
             for (Class<?> parent : getHierarchy(mappingMetadata.getInterfaceType())) {
                 reflectiveClasses.produce(ReflectiveClassBuildItem.builder(parent).methods().build());
             }
 
-            generatedClassesNames.add(mappingMetadata.getClassName());
+            processProperties(mappingMetadata.getInterfaceType(), reflectiveClasses);
+        });
 
-            ClassInfo mappingInfo = combinedIndex.getIndex()
-                    .getClassByName(DotName.createSimple(mappingMetadata.getInterfaceType().getName()));
-            if (mappingInfo != null) {
-                mappingsInfo.add(mappingInfo);
+        configClasses.produce(new ConfigClassBuildItem(configClass, collectTypes(combinedIndex, configClass),
+                generatedClassesNames, prefix, configClassKind));
+    }
+
+    private static void processProperties(
+            Class<?> configClass,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+
+        ConfigMappingInterface mapping = ConfigMappingLoader.getConfigMapping(configClass);
+        for (Property property : mapping.getProperties()) {
+            Class<?> returnType = property.getMethod().getReturnType();
+            reflectiveClasses.produce(ReflectiveClassBuildItem.builder(returnType).methods().build());
+
+            if (property.hasConvertWith()) {
+                Class<? extends Converter<?>> convertWith;
+                if (property.isLeaf()) {
+                    convertWith = property.asLeaf().getConvertWith();
+                } else {
+                    convertWith = property.asPrimitive().getConvertWith();
+                }
+                reflectiveClasses.produce(ReflectiveClassBuildItem.builder(convertWith).build());
```

Comment:
@radcortez how come you don't need `methods` here? What's the purpose of registering the converter if we can't call the `convert` method? What am I missing?

Reply:
The Config system creates by reflection single instances of each `Converter` and calls `convert` as required directly into the instance.

Reply:
> The Config system creates by reflection single instances of each Converter

OK, so that's why we need to register the classes, in order to get them to work with reflection and be able to instantiate them (through the default constructors that are registered by default when creating a `ReflectiveClassBuildItem`).

> and calls convert as required directly into the instance.

Now that's the part I don't get how it works. Since we are registering the class without invoking `methods()` we are not asking GraalVM to register `convert` for reflection as well. So I would expect the call to fail :/

Reply:
But we don't call `convert` by reflection, we lookup the `Converter` type:
https://github.com/smallrye/smallrye-config/blob/c3fcadb5408db6e4b6a356c0bbc88f28af269106/implementation/src/main/java/io/smallrye/config/ConfigMappingContext.java#L103-L124

And call `convert` directly. As an example:
https://github.com/smallrye/smallrye-config/blob/c3fcadb5408db6e4b6a356c0bbc88f28af269106/implementation/src/main/java/io/smallrye/config/ConfigMappingContext.java#L460-L479

Are you having an issue with this?

Reply:
Interesting.

> Are you having an issue with this?

No, I am just trying to understand what's going on.

My curiosity comes from the fact that GraalVM doesn't know if a converter let's say `MyConverter` is reachable, thus treating it as unreachable (otherwise we would have an open world). Now when we register `MyConverter` for reflection we make it reachable, but not its methods (since we don't register the methods as well) which, as a result, I would expect to be left out of the native executable, but apparently they don't.

Thanks for the clarification @radcortez